infrastructure/ansible/roles/nginx/tasks/main.yml

```diff
@@ -4,52 +4,108 @@
+    creates: "{{ dhparams_pem_file }}"
 
 - name: set permissions of dhparam.pem
   become: true
   file:
-    dest: "{{ nginx_ssl_dhparams }}"
+    dest: "{{ dhparams_pem_file }}"
     mode: 600
 
-- name: check for nginx cert key
+
+- name: check for SSL cert key
   stat:
-    path: "{{ nginx_ssl_certificate_key }}"
-  register: cert_key
+    path: "{{ ssl_cert_key }}"
+  register: has_cert_key
   changed_when: false
-
-- name: check for nginx certificate
+ 
+- name: check for SSL cert
   stat:
-    path: "{{ nginx_ssl_certificate }}"
-  register: cert
+    path: "{{ ssl_cert }}"
+  register: has_cert
   changed_when: false
 
 - name: create SSL keys if needed
   become: true
-  when: (cert_key.stat.exists == false) or (cert.stat.exists == false)
-  command: openssl req -x509 -nodes -days 365 -newkey rsa:4096 -keyout {{ nginx_ssl_certificate_key }} -out {{ nginx_ssl_certificate }} -batch -subj '/CN=localhost' -sha256 -addext "subjectAltName = IP:127.0.0.1, IP:::1"
+  when: (has_cert_key.stat.exists == false) or (has_cert.stat.exists == false)
+  command: |
+       openssl req -x509 -nodes -days 365 \
+          -newkey rsa:4096 -keyout {{ ssl_cert_key }} -out {{ ssl_cert }} \
+          -batch -subj '/CN=localhost' -sha256 -addext "subjectAltName = IP:127.0.0.1, IP:::1"
+
+
+- name: check for letsencrypt cert key
+  stat:
+    path: "{{ lets_encrypt_cert_key }}"
+  register: has_lets_encrypt_cert_key
+  changed_when: false
+
+- name: set cert-key variable to generated cert if letsencrypt cert key DNE
+  when: has_lets_encrypt_cert_key.stat.exists == false
+  set_fact:
+    cert_key: "{{ ssl_cert_key }};"
+  changed_when: false
+
+- name: set cert-key variable to lets_encrypt cert-key if exists
+  when: has_lets_encrypt_cert_key.stat.exists
+  set_fact:
+    cert_key: "{{ lets_encrypt_cert_key }}; # managed by Certbot;"
+  changed_when: false
+
+
+- name: check for lets_encrypt cert
+  stat:
+    path: "{{ lets_encrypt_cert }}"
+  register: has_lets_encrypt_cert
+  changed_when: false
+  
+- name: set cert variable to generated cert if lets_encrypt cert DNE
+  when: has_lets_encrypt_cert.stat.exists == false
+  set_fact:
+    cert: "{{ ssl_cert }};"
+  changed_when: false
+
+- name: set cert variable to lets_encrypt cert if exists
+  when: has_lets_encrypt_cert.stat.exists
+  set_fact:
+    cert: "{{ lets_encrypt_cert }}; # managed by Certbot;"
+  changed_when: false
+
 
 - name: deploy nginx sites_enabled configuation
   become: true
   template:
     src: etc_nginx_sites_enabled_default.j2
     dest: /etc/nginx/sites-enabled/default
+  register: nginx_config_changed
 
 - name: allow ports
   become: true
+  tags: firewall
   ufw:
     rule: allow
     port: "{{ item }}"
     proto: tcp
   with_items: "{{ nginx_allowed_ports }}"
 
-- name: relaunch nginx
+- name: restart nginx if config changed
+  when: |
+    (nginx_config_changed.changed) 
+      or (has_cert_key.stat.exists == false) 
+      or (has_cert.stat.exists == false)
```

Comment:
How often is this file run?
Could it happen that the certificate runs out in-between runs?
For Marti and the forum I solved this problem by adding a post-hook that runs `sudo systemctl reload nginx` to ensure that nginx picks up the certificate correctly after refreshing it. It could be the case that certbot does this automatically using the nginx plugin (which is used), but when I initially configured everything there such a plugin didn't exist yer.

Reply:
I just double checked my assumption.
Looks like the nginx plugin takes care of automatically restarting the service, so no need to restart nginx "manually".
See https://community.letsencrypt.org/t/doest-certbot-automaticly-reload-nginx-on-successful-renew/40361/5

Reply:
Is this file `tasks/main.yml`? If so, on every build. If 'this file' is the cert, if you scroll up in the same `tasks/main.yml` there is a stat command to check for `/etc/nginx/cert.crt` and key and generates if does not exist, so that is one-time. 

To explain this when clause, if someone deletes the 'crt' files and we recreate them, and even if 'sites-enabled/default' is not updated, we'll restart nginx. Otherwise, if 'sites-enabled/default' is modified, we'll restart nginx.

> Could it happen that the certificate runs out in-between runs?

This role/task is independent of the certbot certificates, it is for a 'base' nginx only and is unaware of certbot. AFAIK the generated certs `/etc/nginx/cert.crt`, do not expire as those are just keys.

> Looks like the nginx plugin takes care of automatically restarting the service, so no need to restart nginx "manually"

I think perhaps there is a disconnect. We are restarting here on configuration change or a change in the nginx keys, there is no certbot involved here yet.

Reply:
Ah, ok makes sense.
Thanks for clarifying